perfkitbenchmarker/providers/openstack/os_network.py

```diff
@@ -11,150 +11,181 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+""""Module containing classes related to OpenStack Networking."""
 
-import logging
+import json
 import threading
-import time
 
+from perfkitbenchmarker import errors
 from perfkitbenchmarker import flags
 from perfkitbenchmarker import network
+from perfkitbenchmarker import vm_util
 from perfkitbenchmarker.providers.openstack import utils
 from perfkitbenchmarker import providers
 
+
+NEUTRON_FLOAT_IP_SHOW_CMD = 'floatingip-show'
+NEUTRON_FLOAT_IP_DELETE_CMD = 'floatingip-delete'
+NOVA_FLOAT_IP_LIST_CMD = 'floating-ip-list'
+OSC_IP_CMD = 'ip'
+OSC_FLOATING_SUBCMD = 'floating'
+OSC_SEC_GROUP_CMD = 'security group'
+OSC_SEC_GROUP_RULE_CMD = 'security group rule'
+
+SC_GROUP_NAME = 'perfkit_sc_group'
+
+ICMP = 'icmp'
+TCP = 'tcp'
+UDP = 'udp'
+
 FLAGS = flags.FLAGS
 
 MAX_PORT = 65535
 
 
 class OpenStackFirewall(network.BaseFirewall):
+  """
+  An object representing OpenStack Firewall based on Secure Groups.
+  """
+
+  CLOUD = providers.OPENSTACK
+
+  def __init__(self):
+    self._lock = threading.Lock()  # Guards security-group rule set
+    self.sec_group_rules_set = set()
+
+    with self._lock:
+      cmd = utils.OpenStackCLICommand(self, OSC_SEC_GROUP_CMD, 'show',
+                                      SC_GROUP_NAME)
+      stdout, stderr, _ = cmd.Issue(suppress_warning=True)
+      if stderr:
+        cmd = utils.OpenStackCLICommand(self, OSC_SEC_GROUP_CMD, 'create',
+                                        SC_GROUP_NAME)
+        cmd.Issue()
+
+  def AllowICMP(self, vm, icmp_type=-1, icmp_code=-1):
+    """Creates a Security Group Rule on the Firewall to allow/disallow
+    ICMP traffic.
+
+    Args:
+      vm: The BaseVirtualMachine object to allow ICMP traffic to.
+      icmp_type: ICMP type to allow. If none given then allows all types.
+      icmp_code: ICMP code to allow. If none given then allows all codes.
     """
-    An object representing OpenStack Firewall based on Secure Groups.
+    if vm.is_static:
+      return
+
+    sec_group_rule = (ICMP, icmp_type, icmp_code, vm.group_id)
+    with self._lock:
+      cmd = utils.OpenStackCLICommand(vm, OSC_SEC_GROUP_RULE_CMD, 'create',
+                                      vm.group_id)
+      cmd.flags['dst-port'] = str(icmp_type)
+      cmd.flags['proto'] = ICMP
+      if sec_group_rule in self.sec_group_rules_set:
```

Comment:
I don't understand. I see why you want to avoid creating the `sec_group_rule`, but the way this function is written, it looks like you create an OpenStackCLICommand object and set some properties on it before you check whether `sec_group_rule` already exists and isn't needed. I'm suggesting that you check whether `sec_group_rule` exists before creating the OpenStackCLICommand, so you can save a bit more work. Will that not work? Am I missing something?